README.md
## Description

Discord bot that gathers external information about PSO2, including emergency
quest schedules and RSS links.

## Installation

### Python
Python 3.5 or later is required, along with 3 additional packages:
```
python -m pip install -U discord.py
python -m pip install -U pytz
python -m pip install -U aiohttp
```
If your system has multiple python version make sure to call the right python version.

### Discord
1. Sign up for a [Discord](https://discordapp.com/) account
2. Navigate to the [developer portal](https://discordapp.com/developers/applications/)
3. Click "Create an application"
4. Optionally give it a name and picture
5. Keep this page open for later config instructions

### Google
1. Sign up for a Google account
2. Open the [Google APIs](https://console.developers.google.com/cloud-resource-manager) page
3. Create a project
4. Enable APIs and Services
5. Search for and enable **Google Calendar** and **URL shortner**
6. Go to the credentials page and keep it open for later config instructions

### This Bot
1. Get this project using git or download its zip from GitHub and extract it
2. Navigate to the project folder and run the bot `python sla_boy.py`
3. Go to the Discord page from earlier, copy the **CLIENT SECRET**, and paste into config.ini after
   `bot_token = `
4. Go to the Google page from earlier, copy the **API key**, and paste into config.ini after
   `google_api_key = `
5. The bot should run now `python sla_boy.py`
6. Go back to the Discord page from earlier, click the OAuth2 tab and select:
    * **bot** in Scope
    * **Send Messages** in Permissions
    * **Read Message History** in Permissions
7. Copy the generated link and give it to server admins who want to invite the bot
8. Have the admin create a channel for the bot and type `;toggle` in it

## Usage

### Operating the bot
Run `python sla_boy.py` or use the loop scripts `loop.bat` for windows or `./loop.sh` on linux. The
default config contains examples of changeable parameters in the bot and restart the bot after
editing it. 

### Commands
Summary of availiable commands for users. For more detailed information and examples, use `help`.
* `help (command)`: Receive a help message about the bot or detailed information about the command
* `find (text)`: List the next few events with names partially matching the text
* `next`: Show the next event
* `toggle`: Enable/Disable the bot's automatic message (requires **Manage Channels** permission)


## License

This project is licensed under the [MIT License](LICENSE).


changelog.txt
v0.39 @ 2020-07-23
    - Migrate to discord.py v1.0, make sure to update library

v0.38 @ 2019-11-14
    - Removed translated RSS links, google service discontinued
    - Fixed bug where fields are squished together

v0.37 @ 2019-01-25
    - Simplify editing a copy of the loop scripts
    
v0.36 @ 2019-01-09
    - Add a README
    - Add infinite looping scripts for the bot
    - Fix times on calendar display as 60m instead of 1h 00m
    - Fix help message not appearing for "toggle" command
    - Fix help message for next still showing search field
    
v0.35 @ 2018-09-12
    - Make alias groups for PSO2 Summary configurable
    - Default config reloads when user config is
    - Document the default config file
    - Adjust display of remaining minutes to match most system clocks
    - Move alert feed message ahead of the bumped RSS
    - Show decimals when remaining time is a single digit minute 
    - Add helpful links to offline message produced by set_offline.py
    - Fix "next" command sometimes showing events that ended
    - Fix bot not exiting easily with keyboard signals
    
    - Config file changes:
        - Add "PSO2 Summary" section to group similar events
        
v0.34 @ 2017-11-01
    - Tweak "next" command
        - Only print events that are ending in 90 minutes
        - Limit extra events to those 90 minutes away from the nearest event
    - Fix "find" command not working
    

v0.33 @ 2017-10-25
    - Add rss field for pso2.jp/players/
    - Add rss field for bumped.org/
    
    - Config file changes:
        Add "bumped RSS" and "PSO2 RSS" sections
            - "max_items" number of RSS entries to parse
            - "update_interval" delay in seconds before refreshing data

v0.32 @ 2017-10-10
    - Prevent errors from repeatedly printing in update loops
    - Combine all channel messages into a single embed based message
    - Add a bot script to edit an offline bot message

    - Config file changes:
        - Add "channel_update_interval", the seconds to refresh the messages
        - Add "embed_color", the hexidecimal color values for the embed
        - Remove "header" options on all sections
        - Remove Clock's "update_interval", now updates when the message does
        - PSO2 Calendar's default increased to 120 seconds
        - PSO2es Calendar's default increased to 120 seconds

v0.31 @ 2017-04-16
    - Implement a channel message for the PSO2es schedule
    - Reduce cases where the PSO2 schedule was counting events separately
    - Create config file template when no config is found

v0.30 @ 2017-03-22
    - Major overhaul in bot functions
        - Channel updates use multiple messages and data sources now
            - Includes a clock for the chosen timezone
            - No longer combines the schedule and unschedule data
                - Uses a separate message for each now
        - No more new messages upon restarting/crashing
            - Except when the bot messages gets burried by other ones
        - Removed all commands except "future", "next", "set_alerts"
            "future" renamed to "find" and counts the # of events for you
            "set_alerts" renamed to "toggle" and no longer takes any options
            "next" no longer searches, use "find" instead
            commands no longer take time zone option

    - Config file changes:
        - Moved default config out of the doc folder
        - Completely overhauled config options, please see default config.
        
    - "usr" folder is no longer used, you may delete it

v0.17 @ 2016-09-24
    - Hotfix update systems not ignoring errors properly

v0.16 @ 2016-09-22
    - Add "tzlist" and "tzinfo" commands to find time zones easier
    - Update systems are less likely to break

v0.15 @ 2016-09-14
    - Fix bug causing bot to stop updating alerts
    - Changes to properly read new unscheduled event format

v0.14 @ 2016-09-13
    - Change alert remaining time format from "MMmin" to "HH:MM"
    - Send "not found" message to "future" command
    - Bot status now displays help text and time for configured time zone
    - Config file changes:
        - "enable_alert" Enable or disable alert announcements to channels

v0.13 @ 2016-09-11
    - Alert system for scheduled and unscheduled events!
    - Add several commands for managing alerts
        - "remove_alerts", "resend", "set_alerts"
    - Add "allrandom" command to print out unscheduled events
    - Config file changes:
        - channels values reversed (ships to include -> ships to exclude)
        - "alert_every" renamed to "resend_every"
            * no longer resends at intervals, only once per nearing event

v0.12 @ 2016-09-04
    - Fix bug causing "print" command to not work.

v0.11 @ 2016-09-03
    - Help messages for the bot and commands.
    - Future command now limits the results when no search is provided to
      prevent the bot from flooding the user's PM box.
    - Announce command is disabled for now
    - The command prefix by default is now ";" instead of "!'
    - Config file changes:
        - Added "cmd_prefix"
        - Added "find_default"

v0.10 @ 2016-08-30
    Initial alpha release

SLA_bot/gcalutil.py
"""Helper for Google Calendar API

Perform basic retrieval and parsing of Google Calendar events into
CalendarEvent classes.
"""
#If more support is needed for the API, use official py package instead

import datetime as dt
import urllib.parse

import SLA_bot.util as ut

class CalendarEvent:
    """Container for basic information in a scheduled event.
    
    Attributes:
        name (str): Name of the event
        start (datetime): Start date and time of the event
        end (datetime): End date and time of the event
    """
    
    def __init__(self, name, start, end):
        """Return a CalendarEvent with the given name, start and end time.
        
        Args:
            name (str): Name of the event
            start (datetime): Start date and time of the event
            end (datetime): End date and time of the event
        """
        self.name = name
        self.start = start
        self.end = end
        
    def __eq__(self, other):
        return (self.name == other.name and self.start == other.start
               and self.end == other.end)
        
    def __hash__(self):
        return hash((self.name, self.start, self.end))
        
    def status(self, reference_time):
        """Return when the event will start, end, or if it already ended.
        
        Args:
            reference_time (datetime): Whether the event will start, end, or
                has ended at the given time. Often the current time is used.
        
        Returns:
            string: Three different kind of strings depending on the time.
            
            If not started, return the time until the event will start in the
                form "##h ##m" (hours, minutes) up to "##w ##d" (weeks, days).
                
            If started but not ended, return the time until the event ends in
                the form "End ##m" (minutes) up to "End ##w" (weeks).
                
            If the event has ended or the event is missing the start or end
                attribute, return "Ended".
        """
        z_tdelta = dt.timedelta(seconds=0)
        s_tdelta = self.start - reference_time if self.start else z_tdelta
        e_tdelta = self.end - reference_time if self.end else z_tdelta
        if s_tdelta > z_tdelta:
            return ut.two_unit_tdelta(s_tdelta)
        if e_tdelta > z_tdelta:
            return 'End ' + ut.one_unit_tdelta(e_tdelta)
        return 'Ended'
        
    def time_range(self, tz):
        """Return a string of the start time to the end time.
        
        Args:
            tz (timezone): timezone used to localize the start and end time.
            
        Returns:
            string: Month Day, Start(hour:minute) ~ End(hour:minute) Timezone
            
            If the start and end time are more than 23 hours apart, the string
                will include the month and date for the end time as well.
                
            If the start or end time are not set, they will be omitted.
            
            If neither times are set, return the empty string.
        """
        if self.start and self.end:
            start = self.start.astimezone(tz).strftime('%b %d, %H:%M')
            #While events lasting 24 hours or more cannot be written clearly
            #without the date, events lasting 23 hours can but also feel like
            #typos. For example, 13:00 ~ 12:30 can seem more likely to be a
            #typo for 13:00 ~ 13:30 than lasting 23 hours and 30 minutes.
            if self.end - self.start <= dt.timedelta(hours=23):
                end = self.end.astimezone(tz).strftime('%H:%M %Z')
            else:
                end = self.end.astimezone(tz).strftime('%b %d, %H:%M %Z')
            return '{} ~ {}'.format(start, end)
        if self.start:
            return self.start.astimezone(tz).strftime('%b %d, %H:%M %Z')
        if self.end:
            return self.end.astimezone(tz).strftime('%b %d, ??:?? ~ %H:%M %Z')
        return ''
        

def gtimefdt(date_time):
    """Return a string in the API's time format from a datetime."""
    return date_time.isoformat()
    
def gtimepdt(time_string):
    """Return a UTC datetime from an API UTC time string."""
    naive = dt.datetime.strptime(time_string, '%Y-%m-%dT%H:%M:%SZ')
    return naive.replace(tzinfo=dt.timezone.utc)

def parse_data(json_data):
    """Return list of CalendarEvents from the JSON data."""
    events = []
    if not json_data:
        return []
    for item in json_data['items']:
        n = item['summary']
        s = gtimepdt(item['start']['dateTime'])
        e = gtimepdt(item['end']['dateTime'])
        new_event = CalendarEvent(n, s, e)
        events.append(new_event)
    return events

def build_get(cal_id, key, end_min=None, start_max=None):
    """Return GET URL to receive the list of a calendar's events.
    
    Args:
        end_min (datetime): Optional, filter out events that end before this
        start_max (datetime): Optional, filter out events that start at or
            after this
    
    Returns:
        string: A URL of a GET for the calendar with the given query params
    """
    query = {}
    query['maxResults'] = '2500'
    query['timeZone'] = 'UTC'
    query['key'] = key
    if end_min:
        query['timeMin'] = gtimefdt(end_min)
    if start_max:
        query['timeMax'] = gtimefdt(start_max)

    url = 'https://www.googleapis.com/calendar/v3/calendars/{}/events?'
    params = urllib.parse.urlencode(query)
    return url.format(urllib.parse.quote_plus(cal_id)) + params
    
def statusfevents(events, reference_time):
    """Return a string of lines of status and name for events in the list."""
    lines = []
    for e in events:
        status = '{:\xa0>7}'.format(e.status(reference_time))
        lines.append('`|{:^9}|` **{}**'.format(status, e.name))
    return '\n'.join(lines)

